### ISSUE TITLE
opt: Crash on LEFT JOIN with SELECT sub-query

### ISSUE THREAD

**Issue by zrsmith92** [cockroachdb/cockroach#37373]
**Describe the problem**

I'm experiencing a consistent crash when executing a LEFT JOIN query with a sub-SELECT query.

**To Reproduce**

Set up the following schema:
```sql
CREATE TABLE elements (
  id VARCHAR(100) NOT NULL,
  parent_id VARCHAR(100),
  display TEXT,

  PRIMARY KEY (id)
);

CREATE TABLE element_properties (
  element_id VARCHAR(100) NOT NULL,
  property VARCHAR(100) NOT NULL,
  value VARCHAR(100),

  CONSTRAINT fk_element  FOREIGN KEY (element_id) REFERENCES elements (id)
);
```

Then execute this query:

```sql
SELECT *
FROM elements e
LEFT JOIN element_properties p
ON e.id = p.element_id
WHERE (
    e.id = 'third' OR
    e.id IN (SELECT e1.parent_id FROM elements e1 WHERE e1.id = 'third')
);
```

Removing either the `LEFT JOIN` or the `SELECT` subquery fixes it.

**Expected behavior**
I expect this to return a result without crashing

**Additional data / screenshots**
[cockroach.log](https://github.com/cockroachdb/cockroach/files/3155008/cockroach.log)

**Environment:**
 - CockroachDB version 2.1.4
 - Server OS: macOS 10.14.4
 - Client app `cockroach sql`

**Comment by RaduBerinde** [comment]
We are applying to `MapFilterIntoJoinRight` rule for `LeftJoin`. Not sure what changed here since 2.1. CC @rytaft

**Comment by rytaft** [comment]
Sorry for the delay -- I can reproduce this on 2.1, so I'll work on fixing this.

**Comment by rytaft** [comment]
This is the same problem as #36137, which was fixed by #36237. That PR was backported to 19.1 but not to 2.1. @andy-kimball do you want to do the backport?

**Comment by andy-kimball** [comment]
Yes, I can backport.

**Comment by rytaft** [comment]
Fixed by #37615
